### ISSUE TITLE
Write CIFS file 'permission denied' (Linux)

### ISSUE THREAD

**Issue by tai-yi** [dotnet/runtime#42790]
Dear guys,

I got an error when I use my application write to a CIFS file folder mounted on host machine.

I tried use same user with shell script `echo 123 > /mnt/debug/a.txt ` success.

After application error, there is a 0kb file in the folder, means the file is created but application cannot flush binary content into the file.

Mount options `mount -t cifs --verbose -o domain=***,username=***,password='***',dir_mode=0777,file_mode=0777,noperm,rw \
   //somehost/folder /mnt`

error log
```
Unhandled exception. System.UnauthorizedAccessException: Access to the path '/mnt/debug/5d256383-48b6-49f3-99b7-d964c4a964f6' is denied.
 ---> System.IO.IOException: Permission denied
   --- End of inner exception stack trace ---
   at System.IO.FileStream.WriteNative(ReadOnlySpan`1 source)
   at System.IO.FileStream.FlushWriteBuffer()
   at System.IO.FileStream.FlushInternalBuffer()
   at System.IO.FileStream.Flush(Boolean flushToDisk)
   at System.IO.FileStream.Flush()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.IO.StreamWriter.Dispose(Boolean disposing)
   at System.IO.TextWriter.Dispose()
   at System.IO.File.WriteAllText(String path, String contents)
   at NasDebugger.Program.Main(String[] args)
Aborted (core dumped)
```

sample code
``` C#
static int Main(string[] args)
        {
            if (args.Length == 0 || string.IsNullOrEmpty(args[0]))
            {
                Console.WriteLine("please set dir");
                return -1;
            }
            var dir = args[0];
            if (!Directory.Exists(dir))
            {
                Console.WriteLine("please set right dir");
                return -1;
            }
            var id = Guid.NewGuid().ToString();
            File.WriteAllText(Path.Combine(dir, id), "123123");
            Console.WriteLine(id);
            return 0;
        }
```

Right now , I only can reproduce it in my datacenter😭.

OS `Linux version 4.12.14-95.54-default (geeko@buildhost) (gcc version 4.8.5 (SUSE Linux) ) #1 SMP Thu Jun 4 12:49:28 UTC 2020 (892ef1f)`
dotnetcore version: 3.1.8

Does anyone have idea about it?

**Comment by carlossanlop** [comment]
It seems you are hitting issue https://github.com/dotnet/runtime/issues/1816, which has been fixed in [5.0 and master](https://github.com/dotnet/runtime/pull/38742) and was also backported to [3.1](https://github.com/dotnet/coreclr/pull/28099). The next servicing release of .NET Core 3.1 should contain the fix for this. It should go into the November servicing release. Or if you can, you could upgrade your application to 5.0 RC1 and the issue should be gone there.

**Comment by carlossanlop** [comment]
On a second thought, the exception is not exactly the same: `Unhandled exception. System.UnauthorizedAccessException: Access to the path '/mnt/debug/5d256383-48b6-49f3-99b7-d964c4a964f6' is denied.`

So let me ask you a couple of questions to make sure we are on the same page, @tai-yi :
- Do you have permission to write in the remote folder?
- Do you have enough disk space in the remote folder to save your file?
- Is your file larger than the available free disk space?

**Comment by wfurt** [comment]
best option how to troubleshoot this IMHO would be running the app under strace e.g. something like `strace -r -f -s 100  -e trace=%file -e trace=%desc myapp` Do not use `dotnet run`: either use `dotnet myappl.dll` or publish self-contained app.

**Comment by tai-yi** [comment]
@carlossanlop 

* Do you have permission to write in the remote folder?

        yes, I can write with shell script 'echo 123 > /mnt/debug/test'

* Do you have enough disk space in the remote folder to save your file?
  
        yes, I can write successfuly with shell script

* Is your file larger than the available free disk space?
  
        no, I just write text `123123` to the file

**Comment by tai-yi** [comment]
### Update

thanks @wfurt, I will try as your suggestion, this result tested by single file application.

I tried use 4 ways to write file.

1. `File.WriteAllText`
2. `new StreamWriter(path)`
3. `new FileStream(path, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None)`
4. `new FileStream(path, FileMode.OpenOrCreate, FileAccess.Write, FileShare.Read)`
   
Only 3 works fine,, so `FileShare` should impact the result.

#### Sample code
``` C#
class Program
{
    static int Main(string[] args)
    {
        if (args.Length == 0 || string.IsNullOrEmpty(args[0]))
        {
            Console.WriteLine("please set dir");
            return -1;
        }
        var dir = args[0];

        if (!Directory.Exists(dir))
        {
            Console.WriteLine("please set and existing dir");
            return -1;
        }
        var id = Guid.NewGuid().ToString();
        var model = 1;
        if (args.Length > 1) {
            if (!int.TryParse(args[1], out model))
                model = 1;
        }
        var path = Path.Combine(dir, id);
        
        switch (model)
        {
            case 1:
                Write1(path);break;
            case 2:
                Write2(path); break;
            case 3:
                Write3(path); break;
            case 4:
                Write4(path); break;
            default:
                return -1;
        }           
        return 0;
    }

    static void Write1(string path)
    {
        File.WriteAllText(path, "123123");
        Console.WriteLine(path);
    }

    static void Write2(string path)
    {
        using var writer = new StreamWriter(path);
        writer.Write("123123");
        Console.WriteLine(path);
    }

    static void Write3(string path)
    {
        using var fs = new FileStream(path, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None);
        using var writer = new StreamWriter(fs);
        writer.Write("123123");
        Console.WriteLine(path);
    }

    static void Write4(string path)
    {
        using var fs = new FileStream(path, FileMode.OpenOrCreate, FileAccess.Write, FileShare.Read);
        using var writer = new StreamWriter(fs);
        writer.Write("123123");
        Console.WriteLine(path);
    }
}
```

#### sample log
```
the-host-name:/home/debugger # ./NasDebugger /mnt/debug/ 1
Unhandled exception. System.UnauthorizedAccessException: Access to the path '/mnt/debug/c5d0bccf-edc8-4b7b-8607-fab879d0da79' is denied.
 ---> System.IO.IOException: Permission denied
   --- End of inner exception stack trace ---
   at System.IO.FileStream.WriteNative(ReadOnlySpan`1 source)
   at System.IO.FileStream.FlushWriteBuffer()
   at System.IO.FileStream.FlushInternalBuffer()
   at System.IO.FileStream.Flush(Boolean flushToDisk)
   at System.IO.FileStream.Flush()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.IO.StreamWriter.Dispose(Boolean disposing)
   at System.IO.TextWriter.Dispose()
   at System.IO.File.WriteAllText(String path, String contents)
   at NasDebugger.Program.Write1(String path)
   at NasDebugger.Program.Main(String[] args)
Aborted (core dumped)
the-host-name:/home/debugger # ./NasDebugger /mnt/debug/ 2
/mnt/debug/5a973cdd-ce73-40ad-b483-507f3843cbce
Unhandled exception. System.UnauthorizedAccessException: Access to the path '/mnt/debug/5a973cdd-ce73-40ad-b483-507f3843cbce' is denied.
 ---> System.IO.IOException: Permission denied
   --- End of inner exception stack trace ---
   at System.IO.FileStream.WriteNative(ReadOnlySpan`1 source)
   at System.IO.FileStream.FlushWriteBuffer()
   at System.IO.FileStream.FlushInternalBuffer()
   at System.IO.FileStream.Flush(Boolean flushToDisk)
   at System.IO.FileStream.Flush()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.IO.StreamWriter.Dispose(Boolean disposing)
   at System.IO.TextWriter.Dispose()
   at NasDebugger.Program.Write2(String path)
   at NasDebugger.Program.Main(String[] args)
Aborted (core dumped)
the-host-name:/home/debugger # ./NasDebugger /mnt/debug/ 3
/mnt/debug/2e23f4c7-ae3d-488d-968a-a98830777305
the-host-name:/home/debugger # ./NasDebugger /mnt/debug/ 4
/mnt/debug/d82757c5-9b77-4ee0-93c5-c140a4c39f8f
Unhandled exception. System.UnauthorizedAccessException: Access to the path '/mnt/debug/d82757c5-9b77-4ee0-93c5-c140a4c39f8f' is denied.
 ---> System.IO.IOException: Permission denied
   --- End of inner exception stack trace ---
   at System.IO.FileStream.WriteNative(ReadOnlySpan`1 source)
   at System.IO.FileStream.FlushWriteBuffer()
   at System.IO.FileStream.FlushInternalBuffer()
   at System.IO.FileStream.Flush(Boolean flushToDisk)
   at System.IO.FileStream.Flush()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.IO.StreamWriter.Dispose(Boolean disposing)
   at System.IO.TextWriter.Dispose()
   at NasDebugger.Program.Write4(String path)
   at NasDebugger.Program.Main(String[] args)
Aborted (core dumped)
```

**Comment by wfurt** [comment]
the point is not to have single application @tai-yi but to get insight to low-level kernel calls. Pick one of the failing scenarios and get the trace (you can also use `-o trace.txt`)

**Comment by tai-yi** [comment]
@wfurt , thanks for your help.

I tried run command `strace -r -f -s 100 -e trace=file -P /mnt/debug/ -o trace2 ./NasDebugger /mnt/debug/ 2`

here is the content in trace2
```
31866      0.000000 stat("/mnt/debug/", {st_mode=S_IFDIR|0777, st_size=0, ...}) = 0
31866      0.084994 --- SIGABRT {si_signo=SIGABRT, si_code=SI_TKILL, si_pid=31866, si_uid=0} ---
31871      0.003428 +++ killed by SIGABRT (core dumped) +++
31873      0.000018 +++ killed by SIGABRT (core dumped) +++
31869      0.000007 +++ killed by SIGABRT (core dumped) +++
31868      0.000006 +++ killed by SIGABRT (core dumped) +++
31867      0.000006 +++ killed by SIGABRT (core dumped) +++
31870      0.003771 +++ killed by SIGABRT (core dumped) +++
31872      0.000010 +++ killed by SIGABRT (core dumped) +++
31874      0.004476 +++ killed by SIGABRT (core dumped) +++
31866      0.000066 +++ killed by SIGABRT (core dumped) +++
```

**Comment by wfurt** [comment]
There is not much. I think you will also need the `-e trace=%desc`. And I assume the `SIGABRT` is from unhandled execution. (it may be better to catch it and print error message to be sure)

**Comment by tai-yi** [comment]
@wfurt , new trace file uploaded.
[trace2.txt](https://github.com/dotnet/runtime/files/5302718/trace2.txt)

**Comment by wfurt** [comment]
The interesting entry is:
```
29766      0.003445 open("/mnt/debug/eda7ab5a-1a3c-4921-a3b8-b33afba840e9", O_WRONLY|O_CREAT|O_CLOEXEC, 0666) = 23
```
so the open call fails with ENFILE errno. You can try to check with `lsof` to see what files are opened. 
https://stackoverflow.com/questions/24862733/difference-between-linux-errno-23-and-linux-errno-24 may also have some hints. 

One can ask why echo works: as you may see from the trace, dotnet opens lot of file to run. So if you are close to limit that may be just enough. My recommendation would be to check ulimit and fs.file-max and try it again. 

I think we can probably also do better job mapping that particular error. (assuming that the run producing the trace still failed with "Access denied")

**Comment by tai-yi** [comment]
`fs.file-max` is 785044

But I can write successfully with code `new FileStream(path, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None);` and the error only occur when wring content into the file mounted by CIFS.

**Comment by wfurt** [comment]
For comparisons, can you share trace from the working case? What do you get from `lsof | wc -l` ?

**Comment by tai-yi** [comment]
@wfurt , thanks for your help.
Result from `lsof | wc -l` is 20727.

Trace from success case.
[trace3.success.log](https://github.com/dotnet/runtime/files/5310173/trace3.success.log)

**Comment by wfurt** [comment]
Hi @tai-yi,
By looking at the successful case I realized I was wrong in previous comments 

```
open("/mnt/debug/33bf0cf3-4f82-476a-8ef5-bdd4ed08fe79", O_WRONLY|O_CREAT|O_CLOEXEC, 0666) = 23
fstat(23, {st_mode=S_IFREG|0777, st_size=0, ...}) = 0
flock(23, LOCK_EX|LOCK_NB)              = 0
lseek(23, 0, SEEK_CUR)                  = 0
fcntl(1, F_DUPFD_CLOEXEC, 0)            = 24
fstat(24, {st_mode=S_IFCHR|0620, st_rdev=makedev(136, 0), ...}) = 0
write(24, "/mnt/debug/33bf0cf3-4f82-476a-8e"..., 48) = 48
write(23, "123123", 6)                  = 6
flock(23, LOCK_UN)                      = 0
close(23)                               = 0
```

the return value for open is not errno (as usually) but the file handle itself.

going back to trace2.txt:
```
29766      0.003445 open("/mnt/debug/eda7ab5a-1a3c-4921-a3b8-b33afba840e9", O_WRONLY|O_CREAT|O_CLOEXEC, 0666) = 23
```
so open was OK but after that neither flock, fstat or write would happen.  

I assume you do not have SELinux enabled, right (sestatus or  selinuxenabled)
Could you please try to handle the exception before exiting? This should not have impact on the result but I'm puzzled why the WriteNative  would fail without visible attempt to write in the trace.

What is your NAS server? Is there any chance you can give it try with 5.0 bits? (just to be sure)

**Comment by tai-yi** [comment]
@wfurt , new trace uploaded,  selinux should not be enabled, there is no `sestatus` or `selinuxenabled` in the server, I can't find `/sys/fs/selinux `.
I have very limited knowledge about NAS, what is 5.0 bit?
NAS server is managed by another team, they just provide credential and share path.

[trace.error-case2.log](https://github.com/dotnet/runtime/files/5310533/trace.error-case2.log)

**Comment by bad-samaritan** [comment]
Hey @tai-yi,
Just wanted to say thank you so much for your comment on this:

TL;DR: *3 worked for me to!*

> ### Update
> thanks @wfurt, I will try as your suggestion, this result tested by single file application.
> 
> I tried use 4 ways to write file.
> 
> 1. `File.WriteAllText`
> 2. `new StreamWriter(path)`
> 3. `new FileStream(path, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None)`
> 4. `new FileStream(path, FileMode.OpenOrCreate, FileAccess.Write, FileShare.Read)`
> 
> *Only 3 works fine,, so `FileShare` should impact the result.*
> 

Man, I was pulling my hair off why a .NET Core Application got:
```
Exception: System.UnauthorizedAccessException: Access to the path '/SMB/mounted/from/NAS/file.xml' is denied. ---> System.IO.IOException: Permission denied
   --- End of inner exception stack trace ---
   at System.IO.FileStream.WriteNative(ReadOnlySpan`1 source)
   at System.IO.FileStream.FlushWriteBuffer()
   at System.IO.FileStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at System.IO.StreamWriter.Dispose(Boolean disposing)
   at System.IO.TextWriter.Dispose()
   at System.Xml.XmlTextWriter.Close()
   at System.Xml.XmlDocument.Save(String filename)
```

Background:
- NET Core 3.1. Application running in Docker container.
- NAS almost constantly with ~99%/100% storage utilization (reported by `df -h`) BUT ~100 GB free all the time (huge NAS, highly utilized).
- SELinux temporarily disabled to take it out of the picture.

Changes:
- CentOS 7 with kernel upgraded from 3.10 to 5.9.
- All packages updated with yum update.
- Docker updated from 18.x to 19.x (containers required recreation).

I started to notice this shortly after changes mentioned above had been done and the server (hosting containers) had been restarted.
Initially I thought _well... I'm screwed_ because I was convinced it was one of the security / docker update that changed something.
So I tried `cap-add=ALL` and other docker magic - it did not help.
What is interesting is that when I entered the container (`docker exec -it bash`) all operations on the network share executed from bash succeeded - read, write, move, copy etc.
So SAMBA on NAS was set up fine, all permissions fine.

Then I stumbled across this issue.
I changed in the Application:
```xmlDocument.Save(filepathToSmbShare);```
to:
```
using (fs = new FileStream(filepathToSmbShare, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None))
{
    xmlDocument.Save(fs);
}
```
and... it worked like a charm!

So I'm not sure if this is because of the security updates or NAS highly utilized but it saved me.
Thanks again!

**Comment by dennis-yemelyanov** [comment]
Hitting the same issue, and this is really annoying since we have a lot of logging and other code that writes files :(

Instead of writing new file content an empty file gets created and the access exception gets thrown. Writing to the same folder using other means works fine. And there is more than enough disk space on both sides (share server and client)

**Comment by jimbobby5** [comment]
I ran into this and wasn't in a position to change all our code that writes files to CIFS shares. I investigated different kernel versions as it appeared to happen after we upgraded the OS on our hosts. It turned out to be introduced in kernel v5.5.1, in fact this commit:  https://github.com/torvalds/linux/commit/d0677992d2af3d65f1c1c21de3323d09d4891537. I managed to work around it by mounting our cifs shares using the `nobrl` option. From https://linux.die.net/man/8/mount.cifs: 

> nobrl
> Do not send byte range lock requests to the server. This is necessary for certain applications that break with cifs style mandatory byte range locks (and most cifs servers do not yet support requesting advisory byte range locks).

After remounting shares including this option the problem went away.

**Comment by mrmess** [comment]
> 
> 
> I ran into this and wasn't in a position to change all our code that writes files to CIFS shares. I investigated different kernel versions as it appeared to happen after we upgraded the OS on our hosts. It turned out to be introduced in kernel v5.5.1, in fact this commit: [torvalds/linux@d067799](https://github.com/torvalds/linux/commit/d0677992d2af3d65f1c1c21de3323d09d4891537). I managed to work around it by mounting our cifs shares using the `nobrl` option. From https://linux.die.net/man/8/mount.cifs:
> 
> > nobrl
> > Do not send byte range lock requests to the server. This is necessary for certain applications that break with cifs style mandatory byte range locks (and most cifs servers do not yet support requesting advisory byte range locks).
> 
> After remounting shares including this option the problem went away.

THANK YOU! THANK YOU! THANK YOU!

I have been struggling with DOTNET commands on CIFS shares for ages. This is the only thing that fixed it for me.

**Comment by forfcw** [comment]
> I ran into this and wasn't in a position to change all our code that writes files to CIFS shares. I investigated different kernel versions as it appeared to happen after we upgraded the OS on our hosts. It turned out to be introduced in kernel v5.5.1, in fact this commit: [torvalds/linux@d067799](https://github.com/torvalds/linux/commit/d0677992d2af3d65f1c1c21de3323d09d4891537). I managed to work around it by mounting our cifs shares using the `nobrl` option. From https://linux.die.net/man/8/mount.cifs:
> 
> > nobrl
> > Do not send byte range lock requests to the server. This is necessary for certain applications that break with cifs style mandatory byte range locks (and most cifs servers do not yet support requesting advisory byte range locks).
> 
> After remounting shares including this option the problem went away.

You definitely save me form death dude!  I have already search for several days. GOD BLESS YOU!

**Comment by jeffhandley** [comment]
@Jozkee This looks very similar to #44546. When we are looking at that issue, we should also look at this one to see if the same fix will apply to both issues.

**Comment by adamsitnik** [comment]
Most likely related: #53182 #54966

**Comment by adamsitnik** [comment]
I currently don't have an access to a CIFS folder mounted on a different machine and I can't properly test a fix that I've been working on. Is there any chance that someone affected by this bug could run the following app with a path to a writeable folder as an argument and share the output with me?

The app creates a file and tries to acquire few different type of locks before writing to it.

<details>

```cmd
dotnet run -- $pathToFolder
```

```xml
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net5.0</TargetFramework>
        <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    </PropertyGroup>
    <ItemGroup>
      <PackageReference Include="Tmds.LibC" Version="0.5.0" />
    </ItemGroup>
</Project>
```

```cs
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;
using Microsoft.Win32.SafeHandles;
using Tmds.Linux;
using static Tmds.Linux.LibC;

namespace FileSystemRepro
{
    class Program
    {
        static void Main(string[] args)
        {
            string directoryPath = args[0];
            string fileName = Guid.NewGuid().ToString();
            string filePath = Path.Combine(directoryPath, fileName);

            try
            {
                CreateNoLocks(filePath);
                File.Delete(filePath);

                CreateFLock(filePath, LockOperations.LOCK_EX);
                File.Delete(filePath);
                
                CreateFLock(filePath, LockOperations.LOCK_SH);
                File.Delete(filePath);
                
                CreateFcntl(filePath, LockOperations.LOCK_EX);
                File.Delete(filePath);
                
                CreateFcntl(filePath, LockOperations.LOCK_SH);
                File.Delete(filePath);
            }
            finally
            {
                if (File.Exists(filePath))
                {
                    File.Delete(filePath);
                }
            }
        }

        private static void CreateNoLocks(string filePath)
        {
            using (SafeFileHandle sfh = CreateNewFile(filePath))
            using (FileStream fs = new FileStream(sfh, FileAccess.ReadWrite, bufferSize: 1, isAsync: false))
            {
                WriteAndReadSimple(fs);
            }
        }
        
        private static void CreateFLock(string filePath, LockOperations lockOperations)
        {
            using (SafeFileHandle sfh = CreateNewFile(filePath))
            using (FileStream fs = new FileStream(sfh, FileAccess.ReadWrite, bufferSize: 1, isAsync: false))
            {
                Console.WriteLine($"Using {lockOperations} flock to lock the file");
                int lockResult = flock(sfh.DangerousGetHandle().ToInt32(), lockOperations);
                if (lockResult == -1)
                {
                    int errorCode = errno;
                    Console.WriteLine($"Locking {filePath} failed with error code {errorCode} and message: {GetErrorMessage(errorCode)}");
                }

                try
                {
                    WriteAndReadSimple(fs);
                }
                finally
                {
                    flock(sfh.DangerousGetHandle().ToInt32(), LockOperations.LOCK_UN);
                }
            }
        }
        
        private static unsafe void CreateFcntl(string filePath, LockOperations lockOperations)
        {
            using (SafeFileHandle sfh = CreateNewFile(filePath))
            using (FileStream fs = new FileStream(sfh, FileAccess.ReadWrite, bufferSize: 1, isAsync: false))
            {
                Console.WriteLine($"Using {lockOperations} fcntl to lock the file");
                flock flock = default;
                flock.l_whence = (short)SEEK_SET;
                flock.l_start = 0;
                flock.l_len = 0; // lock entire file
                flock.l_type = (short) lockOperations;
                
                int lockResult = fcntl(sfh.DangerousGetHandle().ToInt32(), F_SETLK, &flock);
                if (lockResult == -1)
                {
                    int errorCode = errno;
                    Console.WriteLine($"Locking {filePath} failed with error code {errorCode} and message: {GetErrorMessage(errorCode)}");
                }

                try
                {
                    WriteAndReadSimple(fs);
                }
                finally
                {
                    flock.l_type = (short) LockOperations.LOCK_UN;
                    fcntl(sfh.DangerousGetHandle().ToInt32(), F_SETLK, &flock);
                }
            }
        }

        private static void WriteAndReadSimple(FileStream fs)
        {
            Console.WriteLine($"Length: {fs.Length}, position: {fs.Position}");
            fs.WriteByte(123);
            Console.WriteLine($"After writing one byte: length: {fs.Length}, position: {fs.Position}");
            fs.Position -= 1;
            int read = fs.ReadByte();
            if (read != 123)
            {
                Console.WriteLine($"Invalid content {read}");
            }
        }

        private static unsafe SafeFileHandle CreateNewFile(string filePath)
        {
            byte[] utf8Path = Encoding.UTF8.GetBytes(filePath);
            fixed (byte* pinnedPath = utf8Path)
            {
                int createNewFileFlags = O_CREAT | O_RDWR | O_EXCL;
                mode_t openPermissions =  S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH; // read and write for all
                int result = open(pinnedPath, createNewFileFlags, openPermissions);
                
                if (result == -1)
                {
                    int errorCode = errno;
                    throw new Exception($"Creating {filePath} failed with error code {errorCode} and message: {GetErrorMessage(errorCode)}");
                }

                return new SafeFileHandle(new IntPtr(result), true);
            }
        }
        
        private static unsafe string GetErrorMessage(int errno)
        {
            int bufferLength = 1024;
            byte* buffer = stackalloc byte[bufferLength];

            int rv = strerror_r(errno, buffer, bufferLength);

            return rv == 0 ? Marshal.PtrToStringAnsi((IntPtr)buffer) : $"errno {errno}";
        }

        private enum LockOperations : int
        {
            LOCK_SH = 1,    /* shared lock */
            LOCK_EX = 2,    /* exclusive lock */
            LOCK_NB = 4,    /* don't block when locking*/
            LOCK_UN = 8,    /* unlock */
        }
        
        [DllImport("libc", SetLastError = true, ExactSpelling = true)]
        private static extern int flock(int fd, LockOperations operation);
    }
}
```

</details>

**Comment by ahazelwood** [comment]
> > I ran into this and wasn't in a position to change all our code that writes files to CIFS shares. I investigated different kernel versions as it appeared to happen after we upgraded the OS on our hosts. It turned out to be introduced in kernel v5.5.1, in fact this commit: [torvalds/linux@d067799](https://github.com/torvalds/linux/commit/d0677992d2af3d65f1c1c21de3323d09d4891537). I managed to work around it by mounting our cifs shares using the `nobrl` option. From https://linux.die.net/man/8/mount.cifs:
> > > nobrl
> > > Do not send byte range lock requests to the server. This is necessary for certain applications that break with cifs style mandatory byte range locks (and most cifs servers do not yet support requesting advisory byte range locks).
> > 
> > 
> > After remounting shares including this option the problem went away.
> 
> THANK YOU! THANK YOU! THANK YOU!
> 
> I have been struggling with DOTNET commands on CIFS shares for ages. This is the only thing that fixed it for me.

I also just rant into this after upgrading some linux hosts with a Azure File Share using the 3.1 runtime.  Saved me a ton of additional time trying to track down the issue.
